drivers/gpio/gpio_shell.c

```diff
@@ -1,187 +1,441 @@
+                                                                                                   \
+	static void node_id##cmd_gpio_pin_get(size_t idx, struct shell_static_entry *entry)        \
+	{                                                                                          \
+		gpio_port_pins_t reserved_mask = GPIO_DT_RESERVED_RANGES_NGPIOS_SHELL(node_id);    \
+		uint8_t line_names_len = DT_PROP_LEN_OR(node_id, gpio_line_names, 0);              \
+                                                                                                   \
+		port_pin_get(reserved_mask, node_id##line_names, line_names_len, idx, entry);      \
+		entry->subcmd = NULL;                                                              \
 	}
 
-	return 0;
+#define IS_GPIO_CTRL_PIN_GET(node_id)                                                              \
+	COND_CODE_1(DT_PROP(node_id, gpio_controller), (GPIO_CTRL_PIN_GET_FN(node_id)), ())
+
+DT_FOREACH_STATUS_OKAY_NODE(IS_GPIO_CTRL_PIN_GET)
+
+#define GPIO_CTRL_LIST_ENTRY(node_id)                                                              \
+	{                                                                                          \
+		.dev = DEVICE_DT_GET(node_id),                                                     \
+		.ngpios = DT_PROP_OR(node_id, ngpios, NGPIOS_UNKNOWN),                             \
+		.reserved_mask = GPIO_DT_RESERVED_RANGES_NGPIOS_SHELL(node_id),                    \
+		.line_names = node_id##line_names,                                                 \
+		.line_names_len = DT_PROP_LEN_OR(node_id, gpio_line_names, 0),                     \
+		.subcmd = &node_id##sub_gpio_pin,                                                  \
+	},
+
+#define IS_GPIO_CTRL_LIST(node_id)                                                                 \
+	COND_CODE_1(DT_PROP(node_id, gpio_controller), (GPIO_CTRL_LIST_ENTRY(node_id)), ())
+
+static const struct gpio_ctrl gpio_list[] = {DT_FOREACH_STATUS_OKAY_NODE(IS_GPIO_CTRL_LIST)};
+
+static const struct gpio_ctrl *get_gpio_ctrl(char *name)
+{
+	const struct device *dev = device_get_binding(name);
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(gpio_list); i++) {
+		if (gpio_list[i].dev == dev) {
+			return &gpio_list[i];
+		}
+	}
+	return NULL;
 }
 
-static int cmd_gpio_get(const struct shell *sh,
-			size_t argc, char **argv)
+int line_cmp(const char *input, const char *line_name)
 {
-	const struct device *dev;
-	uint8_t index = 0U;
-	int rc;
+	int i = 0;
 
-	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0) {
-		index = (uint8_t)atoi(argv[args_indx.index]);
-	} else {
-		shell_error(sh, "Wrong parameters for get");
+	while (true) {
+		if ((input[i] == '_') && (line_name[i] == ' ')) {
+			/* Allow underscore to match space */
+		} else if (input[i] != line_name[i]) {
+			return (input[i] > line_name[i]) ? 1 : -1;
+		} else if (line_name[i] == '\0') {
```

Comment:
This check happens in the logic. If the strings are different at any stage, i.e. `input[i]=='\0'` and `line_name[i]` doesn't it will exit otherwise if the characters are equal and `line_name[i]=='\0'` it is a match.